### PR TITLE
Chore: Migrate search and localSearch methods to Typescript

### PR DESCRIPTION
**Pull request by GleidsonDaniel**
## Proposed changes

Migrate search and localSearch methods to Typescript
t
## How to test or reproduce

It is necessary to test all search fields, such as new conversation and room search.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves a current function)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Documentation update (if none of the other choices apply)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat) doc
- [x] I have signed the [CLA](https://cla-assistant.io/RocketChat/Rocket.Chat.ReactNative)
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] I have added necessary documentation (if applicable)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/lib/rocketchat/methods/search.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+			if (filterUsers) {
+				users
+					.filter((item1, index) => users.findIndex(item2 => item2._id === item1._id) === index) // Remove duplicated data from response
+					.filter(user => !data.some(sub => user.username === sub.name)) // Make sure to remove users already on local database
+					.forEach(user => {
+						data.push({
+							...user,
+							rid: user.username,
+							name: user.username,
+							t: SubscriptionType.R_THREAD,
```
Comment (diegolmello) [inline app/lib/rocketchat/methods/search.ts:39]: This is actually `'d'`, so you should use SubscriptionType.DIRECT.

We're just mapping the users returned from the search and adding the type manually to make things easier.

File: app/lib/rocketchat/methods/search.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+import { sanitizeLikeString } from '../../database/utils';
+import database from '../../database/index';
+import { spotlight } from '../services/restApi';
+import isGroupChat from './isGroupChat';
+import { ISearch, ISearchLocal, SubscriptionType } from '../../../definitions';
+
+let debounce: null | ((reason: string) => void) = null;
+
+export const search = async ({ text = '', filterUsers = true, filterRooms = true }): Promise<(ISearch | ISearchLocal)[]> => {
```
Comment (diegolmello) [inline app/lib/rocketchat/methods/search.ts:11]: We should declare `localSearch` before `search` 🙈

File: app/views/NewMessageView.tsx [hunk 3]
```diff
@@ -313,7 +301,7 @@ class NewMessageView extends React.Component<INewMessageViewProps, INewMessageVi
 			<FlatList
 				data={search.length > 0 ? search : chats}
 				extraData={this.state}
-				keyExtractor={item => item._id}
+				keyExtractor={item => item._id || item.rid}
```
Comment (diegolmello) [inline app/views/NewMessageView.tsx:304]: Now with the code more readable, I think we can remove that `_id`, can't we?

**Review by diegolmello — CHANGES_REQUESTED** [review]

File: app/views/NewMessageView.tsx (hunk 3 above)
Comment (GleidsonDaniel) [reply app/views/NewMessageView.tsx:304]: Worse than not, when the object comes directly from the api it doesn't have the rid.

**Review by diegolmello — APPROVED** [review]

**Pull request merged**